File: Python/Leetcode 2558.py
import math
from queue import PriorityQueue

class Solution:
    def pickGifts(self, gifts: List[int], k: int) -> int:
        
        pq = PriorityQueue()
        
        for number in gifts:
            pq.put(-1 * number)
        
        while k:
            k = k - 1
            
            top  = -1*int(pq.get())
            print(top , math.sqrt(top))
            pq.put(-1*math.sqrt(top))
            
        
        answer = 0
        print(pq.qsize())
        while not pq.empty():
            answer = answer + -1*int(pq.get())
            
        return answer
        
        

File: C++/Leetcode 2558.cpp
#include <cmath>
#include <stdio.h>
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    long long pickGifts(vector<int>& gifts, int k) {
        priority_queue<int> pq;
        
        for (int i = 0; i < gifts.size(); i++)
        {
            pq.push(gifts[i]);
        }
        
        while (k)
        {
            k--;
            
            int top = pq.top();
            
            pq.pop();
            
            long long sqrt = pow(top,0.5);
            
            pq.push(sqrt);
        }
        
        long long answer = 0;
        while (pq.size())
        {
            answer += pq.top();
            pq.pop();
        }
        
        return answer;
    }
};